### DISCUSSION TITLE
Binding Image

### DISCUSSION THREAD

**Discussion by CarlosEX** [AvaloniaUI/Avalonia#5558] [Q&A] [answered]
Friends, I'm not able to understand and make the dynamic link of images. Where I'm wrong, or what's missing.

I put the test project on my GitHub: https://github.com/CarlosEX/ExampleBindingImage

------------------------------------------------------

# BitmapAssetValueConverter.cs

This is my converter class i picked up from the documentation it's inside my ViewModels folder

```
using System;
using System.Globalization;
using System.Reflection;
using Avalonia;
using Avalonia.Data.Converters;
using Avalonia.Media.Imaging;
using Avalonia.Platform;

namespace TestBindingImage.ViewModels
{
    public class BitmapAssetValueConverter : IValueConverter
    {
        public static BitmapAssetValueConverter Instance = new BitmapAssetValueConverter();

        public object? Convert(object value, Type targetType, object parameter, CultureInfo culture)
        {
            if (value == null)
                return null;

            if (value is string rawUri && targetType == typeof(IBitmap))
            {
                Uri uri;

                // Allow for assembly overrides
                if (rawUri.StartsWith("avares://"))
                {
                    uri = new Uri(rawUri);
                }
                else
                {
                    string? assemblyName = Assembly.GetEntryAssembly()?.GetName().Name;
                    uri = new Uri($"avares://{assemblyName}{rawUri}");
                }

                var assets = AvaloniaLocator.Current.GetService<IAssetLoader>();
                var asset = assets.Open(uri);

                return new Bitmap(asset);
            }
            throw new NotSupportedException();
        }

        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
        {
            throw new NotSupportedException();
        }
    }
}
```
--------------------------------------------------------------------
# MainWindowViewModel.cs

```

namespace TestBindingImage.ViewModels
{
    public class MainWindowViewModel : ViewModelBase
    {
        public string PathImage => "/Assets/avalonia-logo.ico";
        
    }
}
```

# MainWindow.axaml

```
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:vm="using:TestBindingImage.ViewModels"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:ext="clr-namespace:TestBindingImage.ViewModels"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="TestBindingImage.Views.MainWindow"
        Icon="/Assets/avalonia-logo.ico"
        Title="TestBindingImage">

    <Design.DataContext>
        <vm:MainWindowViewModel/>
    </Design.DataContext>

    <Window.Resources>
        <ext:BitmapAssetValueConverter x:Key="variableImage"/>
    </Window.Resources>

    <Grid>
        <Image Width="75" Height="73" Source="{Binding PathImage, Converter={StaticResource variableImage}}"></Image>
    </Grid>

</Window>

```

**Comment by jp2masa** [comment, marked as answer]
https://github.com/AvaloniaUI/avaloniaui.net/issues/235
https://github.com/AvaloniaUI/avaloniaui.net/pull/237

**Reply by CarlosEX** [reply]
Thank you very much, 24 hours banging head to solve a problem that was only replace *IBitmap* by *IImage* . But it was good, together came other knowledge.

**Reply by CarlosEX** [reply]
@jp2masa  It worked with the example of the ico image that already comes by default of avalonia. But I switched to a png and jpg image and it didn't work. Do I need to set something up on my converter?

**Reply by maxkatz6** [reply]
No, png/jpg works with this approach, if they are added in the csproj as a avalonia resource.
Do not forget to provide additional information like exception information, etc.

**Reply by jp2masa** [reply]
It should work without changes to the converter. Do you have a repro?

**Reply by CarlosEX** [reply]
I tried these here but none worked with png or jpg
1. if (value is string rawUri && targetType.IsAssignableFrom(typeof(IImage)))
2. if (value is string rawUri && targetType.IsAssignableFrom(typeof(IBitmap)))
3. if (value is string rawUri && targetType.IsAssignableFrom(typeof(IBitmapImpl)))
4. if (value is string rawUri && targetType == typeof(IImage))
5. if (value is string rawUri && targetType == typeof(IBitmap))
6. if (value is string rawUri && targetType == typeof(Bitmap))


I created this sample repository @jp2masa @maxkatz6 : https://github.com/CarlosEX/ExampleBindingImage

**Reply by jp2masa** [reply]
It works for me, did you try to rebuild the project?

**Reply by CarlosEX** [reply]
@jp2masa Yes, I've already compiled. But you're just picking on the .ico but I believe I'm the one who's missing something out of limitation of knowledge.

**Reply by jp2masa** [reply]
I was talking about the png in your repo, it was displayed correctly for me.

**Reply by CarlosEX** [reply]
I get it, I really don't know what's going on on my computer. I'm going to keep digging here. Still thank you so much for the help.

**Reply by CarlosEX** [reply]
@jp2masa @maxkatz6 Now it's gone. Do you know what it was? The definitions of Heigth and Width that were interfering. When I took it off, it worked. Showww

**Comment by jaromir-dobes** [comment]
Is there any special setup required for unit testing of this code? I'm getting NullReferenceException from the Bitmap constructor if I call Convert method in an xunit test.

**Reply by maxkatz6** [reply]
What are you using for unit tests? Avalonia.Headless ?

**Reply by maxkatz6** [reply]
I think it's better to create another issue with more details
